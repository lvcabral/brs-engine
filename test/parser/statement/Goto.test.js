const brs = require("../../../bin/brs.node");
const { token, identifier, EOF } = require("../ParserTests");
const { Lexeme } = brs.lexer;

describe("parser goto statements", () => {
    it("parses standalone statement properly", () => {
        let { statements, errors } = brs.parser.Parser.parse([
            token(Lexeme.Goto, "goto"),
            identifier("SomeLabel"),
            EOF,
        ]);
        expect(errors.length).toEqual(0);
        expect({ errors, statements }).toMatchSnapshot();
    });

    it("detects labels", () => {
        let { statements, errors } = brs.parser.Parser.parse([
            identifier("SomeLabel"),
            token(Lexeme.Colon, ":"),
            EOF,
        ]);
        expect(errors.length).toEqual(0);
        expect(statements).toMatchSnapshot();
    });

    it("allows multiple goto statements on one line", () => {
        let { tokens } = brs.lexer.Lexer.scan(`
            sub Main()
                'multiple goto statements on one line
                goto myLabel : goto myLabel
                myLabel:
            end sub
        `);
        let { statements, errors } = brs.parser.Parser.parse(tokens);
        expect(errors.length).toEqual(0);
        expect(statements).toMatchSnapshot();
    });

    it("do not allow label inside try catch", () => {
        let { tokens } = brs.lexer.Lexer.scan(`
            sub main()
                goto inside
                try
                    inside:
                    throw "something"
                catch ex
                    print ex
                end try
            end sub
        `);
        let { statements, errors } = brs.parser.Parser.parse(tokens);
        expect(errors.length).toEqual(1);
        expect(errors[0].message).toEqual(
            "Syntax Error. (compile error &h2) Labels are illegal inside a TRY clause."
        );
    });
});
