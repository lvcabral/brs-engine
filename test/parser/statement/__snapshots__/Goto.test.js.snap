// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser goto statements allows multiple goto statements on one line 1`] = `
[
  Function {
    "func": Function {
      "body": Block {
        "location": {
          "end": {
            "column": 12,
            "line": 6,
          },
          "file": "",
          "start": {
            "column": 22,
            "line": 2,
          },
        },
        "statements": [
          Goto {
            "tokens": {
              "goto": {
                "isReserved": true,
                "kind": "Goto",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 20,
                    "line": 4,
                  },
                  "file": "",
                  "start": {
                    "column": 16,
                    "line": 4,
                  },
                },
                "text": "goto",
              },
              "label": {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 28,
                    "line": 4,
                  },
                  "file": "",
                  "start": {
                    "column": 21,
                    "line": 4,
                  },
                },
                "text": "myLabel",
              },
            },
          },
          Goto {
            "tokens": {
              "goto": {
                "isReserved": true,
                "kind": "Goto",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 35,
                    "line": 4,
                  },
                  "file": "",
                  "start": {
                    "column": 31,
                    "line": 4,
                  },
                },
                "text": "goto",
              },
              "label": {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 43,
                    "line": 4,
                  },
                  "file": "",
                  "start": {
                    "column": 36,
                    "line": 4,
                  },
                },
                "text": "myLabel",
              },
            },
          },
          Label {
            "tokens": {
              "colon": {
                "isReserved": false,
                "kind": "Colon",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 24,
                    "line": 5,
                  },
                  "file": "",
                  "start": {
                    "column": 23,
                    "line": 5,
                  },
                },
                "text": ":",
              },
              "identifier": {
                "isReserved": false,
                "kind": "Identifier",
                "literal": undefined,
                "location": {
                  "end": {
                    "column": 23,
                    "line": 5,
                  },
                  "file": "",
                  "start": {
                    "column": 16,
                    "line": 5,
                  },
                },
                "text": "myLabel",
              },
            },
          },
        ],
      },
      "end": {
        "isReserved": false,
        "kind": "EndSub",
        "literal": undefined,
        "location": {
          "end": {
            "column": 19,
            "line": 6,
          },
          "file": "",
          "start": {
            "column": 12,
            "line": 6,
          },
        },
        "text": "end sub",
      },
      "keyword": {
        "isReserved": true,
        "kind": "Sub",
        "literal": undefined,
        "location": {
          "end": {
            "column": 15,
            "line": 2,
          },
          "file": "",
          "start": {
            "column": 12,
            "line": 2,
          },
        },
        "text": "sub",
      },
      "parameters": [],
      "returns": 11,
    },
    "name": {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": {
        "end": {
          "column": 20,
          "line": 2,
        },
        "file": "",
        "start": {
          "column": 16,
          "line": 2,
        },
      },
      "text": "Main",
    },
  },
]
`;

exports[`parser goto statements detects labels 1`] = `
[
  Label {
    "tokens": {
      "colon": {
        "isReserved": false,
        "kind": "Colon",
        "literal": undefined,
        "location": {
          "end": {
            "column": -9,
            "line": -9,
          },
          "start": {
            "column": -9,
            "line": -9,
          },
        },
        "text": ":",
      },
      "identifier": {
        "isReserved": false,
        "kind": "Identifier",
        "literal": undefined,
        "location": {
          "end": {
            "column": -9,
            "line": -9,
          },
          "start": {
            "column": -9,
            "line": -9,
          },
        },
        "text": "SomeLabel",
      },
    },
  },
]
`;

exports[`parser goto statements parses standalone statement properly 1`] = `
{
  "errors": [],
  "statements": [
    Goto {
      "tokens": {
        "goto": {
          "isReserved": true,
          "kind": "Goto",
          "literal": undefined,
          "location": {
            "end": {
              "column": -9,
              "line": -9,
            },
            "start": {
              "column": -9,
              "line": -9,
            },
          },
          "text": "goto",
        },
        "label": {
          "isReserved": false,
          "kind": "Identifier",
          "literal": undefined,
          "location": {
            "end": {
              "column": -9,
              "line": -9,
            },
            "start": {
              "column": -9,
              "line": -9,
            },
          },
          "text": "SomeLabel",
        },
      },
    },
  ],
}
`;
